File: packages/python/elizaos/bootstrap/actions/image_generation.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.bootstrap.utils.xml import parse_key_value_xml
from elizaos.prompts import IMAGE_GENERATION_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class GenerateImageAction:
    name: str = "GENERATE_IMAGE"
    similes: list[str] = field(
        default_factory=lambda: [
            "CREATE_IMAGE",
            "MAKE_IMAGE",
            "DRAW",
            "PAINT",
            "VISUALIZE",
            "RENDER_IMAGE",
        ]
    )
    description: str = (
        "Generate an image using AI image generation models. "
        "Use this when the user requests visual content or imagery."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        return runtime.has_model(ModelType.IMAGE)

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for GENERATE_IMAGE action")

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ACTION_STATE"])

        template = (
            runtime.character.templates.get("imageGenerationTemplate")
            if runtime.character.templates
            and "imageGenerationTemplate" in runtime.character.templates
            else IMAGE_GENERATION_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)

        prompt_response = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(prompt_response)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response for image prompt")

        thought = str(parsed_xml.get("thought", ""))
        image_prompt = str(parsed_xml.get("prompt", ""))

        if not image_prompt:
            raise ValueError("No image prompt generated")

        image_result = await runtime.use_model(
            ModelType.IMAGE,
            prompt=image_prompt,
        )

        image_url: str | None = None
        if isinstance(image_result, str):
            image_url = image_result
        elif isinstance(image_result, dict):
            image_url = image_result.get("url") or image_result.get("data")

        if not image_url:
            raise ValueError("No image URL returned from generation")

        response_content = Content(
            text=f"Generated image with prompt: {image_prompt}",
            attachments=[{"type": "image", "url": image_url}],
            actions=["GENERATE_IMAGE"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Generated image: {image_prompt}",
            values={
                "success": True,
                "imageGenerated": True,
                "imageUrl": image_url,
                "imagePrompt": image_prompt,
            },
            data={
                "actionName": "GENERATE_IMAGE",
                "prompt": image_prompt,
                "thought": thought,
                "imageUrl": image_url,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Can you draw a sunset over the ocean?"),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll generate that image for you.",
                        actions=["GENERATE_IMAGE"],
                    ),
                ),
            ],
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Create an image of a futuristic city."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="Generating a futuristic cityscape...",
                        actions=["GENERATE_IMAGE"],
                    ),
                ),
            ],
        ]


generate_image_action = Action(
    name=GenerateImageAction.name,
    similes=GenerateImageAction().similes,
    description=GenerateImageAction.description,
    validate=GenerateImageAction().validate,
    handler=GenerateImageAction().handler,
    examples=GenerateImageAction().examples,
)

File: packages/typescript/src/bootstrap/actions/imageGeneration.ts
import { v4 } from "uuid";
import { logger } from "../../logger.ts";
import { imageGenerationTemplate } from "../../prompts.ts";
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "../../types/index.ts";
import { ContentType, ModelType } from "../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../utils.ts";

export const generateImageAction = {
  name: "GENERATE_IMAGE",
  similes: ["DRAW", "CREATE_IMAGE", "RENDER_IMAGE", "VISUALIZE"],
  description:
    "Generates an image based on a generated prompt reflecting the current conversation. Use GENERATE_IMAGE when the agent needs to visualize, illustrate, or demonstrate something visually for the user.",
  validate: async (_runtime: IAgentRuntime) => {
    return true;
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    responses?: Memory[],
  ): Promise<ActionResult> => {
    const allProviders =
      responses?.flatMap((res) => res.content?.providers || []) || [];

    state = await runtime.composeState(message, [
      ...(allProviders ?? []),
      "RECENT_MESSAGES",
    ]);

    const prompt = composePromptFromState({
      state,
      template:
        runtime.character.templates?.imageGenerationTemplate ||
        imageGenerationTemplate,
    });

    const promptResponse = await runtime.useModel(ModelType.TEXT_LARGE, {
      prompt,
    });

    const parsedXml = parseKeyValueXml(promptResponse);
    const promptValue = parsedXml?.prompt;

    const imagePrompt: string =
      typeof promptValue === "string"
        ? promptValue
        : "Unable to generate descriptive prompt for image";

    const imageResponse = await runtime.useModel(ModelType.IMAGE, {
      prompt: imagePrompt,
    });

    if (
      !imageResponse ||
      imageResponse.length === 0 ||
      !imageResponse[0]?.url
    ) {
      logger.error(
        {
          src: "plugin:bootstrap:action:image_generation",
          agentId: runtime.agentId,
          imagePrompt,
        },
        "Image generation failed - no valid response received",
      );
      return {
        text: "Image generation failed",
        values: {
          success: false,
          error: "IMAGE_GENERATION_FAILED",
          prompt: imagePrompt,
        },
        data: {
          actionName: "GENERATE_IMAGE",
          prompt: imagePrompt,
          rawResponse: imageResponse,
        },
        success: false,
      };
    }

    const imageUrl = imageResponse[0].url;

    logger.info(
      {
        src: "plugin:bootstrap:action:image_generation",
        agentId: runtime.agentId,
        imageUrl,
      },
      "Received image URL",
    );

    const getFileExtension = (url: string): string => {
      const urlPath = new URL(url).pathname;
      const urlPathSplit = urlPath.split(".");
      const extension =
        urlPathSplit.length > 0 &&
        urlPathSplit[urlPathSplit.length - 1] &&
        urlPathSplit[urlPathSplit.length - 1].toLowerCase();
      if (
        extension &&
        ["png", "jpg", "jpeg", "gif", "webp", "bmp"].includes(extension)
      ) {
        return extension;
      }
      return "png";
    };

    const extension = getFileExtension(imageUrl);
    const timestamp = new Date()
      .toISOString()
      .replace(/[:.]/g, "-")
      .slice(0, 19);
    const fileName = `Generated_Image_${timestamp}.${extension}`;
    const attachmentId = v4();

    const responseContent = {
      attachments: [
        {
          id: attachmentId,
          url: imageUrl,
          title: fileName,
          contentType: ContentType.IMAGE,
        },
      ],
      thought: `Generated an image based on: "${imagePrompt}"`,
      actions: ["GENERATE_IMAGE"],
      text: imagePrompt,
    };

    if (callback) {
      await callback(responseContent);
    }

    return {
      text: "Generated image",
      values: {
        success: true,
        imageGenerated: true,
        imageUrl,
        prompt: imagePrompt,
      },
      data: {
        actionName: "GENERATE_IMAGE",
        imageUrl,
        prompt: imagePrompt,
      },
      success: true,
    };
  },
  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Can you show me what a futuristic city looks like?",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Sure, I'll create a futuristic city image for you. One moment...",
          actions: ["GENERATE_IMAGE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "What does a neural network look like visually?",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "I’ll create a visualization of a neural network for you, one sec...",
          actions: ["GENERATE_IMAGE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "Can you visualize the feeling of calmness for me?",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Creating an image to capture calmness for you, please wait a moment...",
          actions: ["GENERATE_IMAGE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "What does excitement look like as an image?",
        },
      },
      {
        name: "{{name2}}",
        content: {
          text: "Let me generate an image that represents excitement for you, give me a second...",
          actions: ["GENERATE_IMAGE"],
        },
      },
    ],
  ] as ActionExample[][],
} as Action;
